drivers/tee/optee/optee.c

```diff
@@ -0,0 +1,1267 @@
+	const char *method;
+};
+
+struct optee_notify {
+	sys_dnode_t node;
+	uint32_t key;
+	struct k_sem wait;
+};
+
+struct optee_supp_req {
+	sys_dnode_t link;
+
+	bool in_queue;
+	uint32_t func;
+	uint32_t ret;
+	size_t num_params;
+	struct tee_param *param;
+
+	struct k_sem complete;
+};
+
+struct optee_supp {
+	/* Serializes access to this struct */
+	struct k_mutex mutex;
+
+	int req_id;
+	sys_dlist_t reqs;
+	struct optee_supp_req *current;
+	struct k_sem reqs_c;
+};
+
+struct optee_driver_data {
+	smc_call_t smc_call;
+
+	sys_bitarray_t *notif_bitmap;
+
+	sys_dlist_t notif;
+	struct k_spinlock notif_lock;
+	struct optee_supp supp;
+	unsigned long sec_caps;
+};
+
+/* Wrapping functions so function pointer can be used */
+static void optee_smccc_smc(unsigned long a0, unsigned long a1, unsigned long a2, unsigned long a3,
+			    unsigned long a4, unsigned long a5, unsigned long a6, unsigned long a7,
+			    struct arm_smccc_res *res)
+{
+	arm_smccc_smc(a0, a1, a2, a3, a4, a5, a6, a7, res);
+}
+
+static void optee_smccc_hvc(unsigned long a0, unsigned long a1, unsigned long a2, unsigned long a3,
+			    unsigned long a4, unsigned long a5, unsigned long a6, unsigned long a7,
+			    struct arm_smccc_res *res)
+{
+	arm_smccc_hvc(a0, a1, a2, a3, a4, a5, a6, a7, res);
+}
+
+static int param_to_msg_param(const struct tee_param *param, unsigned int num_param,
+			      struct optee_msg_param *msg_param)
+{
+	int i;
+	const struct tee_param *tp = param;
+	struct optee_msg_param *mtp = msg_param;
+
+	if (!param || !msg_param) {
+		return -EINVAL;
+	}
+
+	for (i = 0; i < num_param; i++, tp++, mtp++) {
+		if (!tp || !mtp) {
+			LOG_ERR("Wrong param on %d iteration", i);
+			return -EINVAL;
+		}
+
+		switch (tp->attr) {
+		case TEE_PARAM_ATTR_TYPE_NONE:
+			mtp->attr = OPTEE_MSG_ATTR_TYPE_NONE;
+			memset(&mtp->u, 0, sizeof(mtp->u));
+			break;
+		case TEE_PARAM_ATTR_TYPE_VALUE_INPUT:
+		case TEE_PARAM_ATTR_TYPE_VALUE_OUTPUT:
+		case TEE_PARAM_ATTR_TYPE_VALUE_INOUT:
+			mtp->attr = OPTEE_MSG_ATTR_TYPE_VALUE_INPUT + tp->attr -
+				TEE_PARAM_ATTR_TYPE_VALUE_INPUT;
+			mtp->u.value.a = tp->a;
+			mtp->u.value.b = tp->b;
+			mtp->u.value.c = tp->c;
+			break;
+		case TEE_PARAM_ATTR_TYPE_MEMREF_INPUT:
+		case TEE_PARAM_ATTR_TYPE_MEMREF_OUTPUT:
+		case TEE_PARAM_ATTR_TYPE_MEMREF_INOUT:
+			mtp->attr = OPTEE_MSG_ATTR_TYPE_RMEM_INPUT + tp->attr -
+				TEE_PARAM_ATTR_TYPE_MEMREF_INPUT;
+			mtp->u.rmem.shm_ref = tp->c;
+			mtp->u.rmem.size = tp->b;
+			mtp->u.rmem.offs = tp->a;
+			break;
+		default:
+			return -EINVAL;
+		}
+	}
+
+	return 0;
+}
+
+static void msg_param_to_tmp_mem(struct tee_param *p, uint32_t attr,
+				 const struct optee_msg_param *mp)
+{
+	struct tee_shm *shm = (struct tee_shm *)mp->u.tmem.shm_ref;
+
+	p->attr = TEE_PARAM_ATTR_TYPE_MEMREF_INPUT + attr - OPTEE_MSG_ATTR_TYPE_TMEM_INPUT;
+	p->b = mp->u.tmem.size;
+
+	if (!shm) {
+		p->a = 0;
+		p->c = 0;
+		return;
+	}
+
+	p->a = mp->u.tmem.buf_ptr - z_mem_phys_addr(shm->addr);
```

Comment:
z_mem_phys_addr was deprecated, please use k_mem_phys_addr instead.

Reply:
thank you. fixed.